livepeer/lpms
# Support H.265/HEVC, VP8, VP9

Issue link: https://github.com/livepeer/go-livepeer/issues/2123

Details:
* Added support for H.265 codecs: libx265 (CPU, encoding and decoding), NVENC (GPU, encoding and decoding)
* Added support for VP8/VP9 codecs: libvpx (CPU, encoding and decoding), NVENC (GPU, decoding only, where supported)
* Added transcoding test for all combinations of codecs and acceleration

Testing:
* install dependencies using [install_ffmpeg.sh](https://github.com/livepeer/go-livepeer/blob/ip/hevc/install_ffmpeg.sh) (WIP, please report errors)
* run tests:
  ```
  go test -v --tags=nvidia -timeout 30m -run TestNvidia_Transcoding ./ffmpeg/
  go test -v --tags=nvidia -timeout 30m -run TestSW_Transcoding ./ffmpeg/
  ```
They will output lines like below, and fail, if anything is not supported (on CI, only CPU test will run for now). SW test should work anywhere, Nvidia test should pass on GPUs starting from desktop GTX 1070 - it can't decode VP9 on my GTX 1070m, despite support matrix stating it should.
```
    ffmpeg_test.go:754: Transcode VP8 (Accel: 0) -> H.265/HEVC (Accel: 0) Prepare: true Transcode: true Quality: true
    ffmpeg_test.go:754: Transcode VP8 (Accel: 0) -> VP8 (Accel: 0) Prepare: true Transcode: true Quality: true
    ffmpeg_test.go:754: Transcode VP8 (Accel: 0) -> VP9 (Accel: 0) Prepare: true Transcode: true Quality: true
```
